src/cli/upgrade.rs

```diff
@@ -0,0 +1,169 @@
+use std::cmp::Ordering;
+use std::str::FromStr;
+
+use crate::cli::cli_config::ProjectConfig;
+use crate::Project;
+use clap::Parser;
+use fancy_display::FancyDisplay;
+use itertools::Itertools;
+use miette::MietteDiagnostic;
+
+use pixi_manifest::Feature;
+use pixi_manifest::FeatureName;
+use pixi_manifest::SpecType;
+use rattler_conda_types::MatchSpec;
+
+use super::cli_config::PrefixUpdateConfig;
+
+/// Update dependencies as recorded in the local lock file
+#[derive(Parser, Debug, Default)]
+pub struct Args {
+    #[clap(flatten)]
+    pub project_config: ProjectConfig,
+
+    #[clap(flatten)]
+    pub prefix_update_config: PrefixUpdateConfig,
+
+    #[clap(flatten)]
+    pub specs: UpgradeSpecsArgs,
+}
+
+#[derive(Parser, Debug, Default)]
+pub struct UpgradeSpecsArgs {
+    /// The packages to upgrade
+    pub packages: Option<Vec<String>>,
+
+    /// The feature to update
+    #[clap(long = "feature", short = 'f', default_value_t)]
+    pub feature: FeatureName,
+}
+
+pub async fn execute(args: Args) -> miette::Result<()> {
+    let mut project = Project::load_or_else_discover(args.project_config.manifest_path.as_deref())?
+        .with_cli_config(args.prefix_update_config.config.clone());
+
+    // Ensure that the given feature exists
+    let Some(feature) = project.manifest.feature(&args.specs.feature) else {
+        miette::bail!(
+            "could not find a feature named {}",
+            args.specs.feature.fancy_display()
+        )
+    };
+
+    // If the user specified a package name, check to see if it is even there.
+    if let Some(packages) = &args.specs.packages {
+        for package in packages {
+            ensure_package_exists(feature, package)?
+        }
+    }
+
+    // TODO: Also support build and host
+    let spec_type = SpecType::Run;
+    let match_specs = feature
+        .dependencies(Some(spec_type), None)
+        .into_iter()
+        .flat_map(|deps| deps.into_owned())
+        .filter(|(name, _)| match &args.specs.packages {
+            None => true,
+            Some(packages) if packages.contains(&name.as_normalized().to_string()) => true,
+            _ => false,
+        })
+        .filter_map(|(name, spec)| {
+            match spec.try_into_nameless_match_spec(&project.channel_config()) {
+                Ok(Some(mut nameless_spec)) => {
+                    // In order to upgrade, we remove the version requirement
+                    nameless_spec.version = None;
+                    Some((
+                        name.clone(),
+                        (
+                            MatchSpec::from_nameless(nameless_spec, Some(name)),
+                            spec_type,
+                        ),
+                    ))
```

Comment:
I think we should parse the input as match_spec and inject it here.

Currently the implementation skips `pixi upgrade "numpy<1.22"` and just installs `numpy` to the latest version.
Also this will not work if the user specified a build dependency for instance. I do think we should copy the `channel` information, and possibly more but it doesn't come to mind yet.